File: js/priorityQ.js
(function () {'use strict';}());

class PriorityQueue{
    constructor(){
        this._priorityDict = {};
    }

    insert(value, priority=0){
        if(this._priorityDict[priority] !== undefined){
            this._priorityDict[priority].push(value)
        }else{
            this._priorityDict[priority] = priority
            this._priorityDict[priority] = []
            this._priorityDict[priority].push(value)
        }
    }

    pop(){
        if(isEmptyObject(this._priorityDict)){
            throw new Error('There is nothing to pop')
        }
        let highestPriority = Math.max.apply(Math, Object.keys(this._priorityDict).map(Number))
        let popped = this._priorityDict[highestPriority].shift()
        if(isEmptyObject(this._priorityDict[highestPriority])){
            delete this._priorityDict[highestPriority]
        }
        return popped
    }

    peek(){
        if(isEmptyObject(this._priorityDict)){
            return undefined
        }
        let highestPriority = Math.max.apply(Math, Object.keys(this._priorityDict).map(Number))
        return this._priorityDict[highestPriority][0]
    }
}

function isEmptyObject( obj ) {
    for ( var name in obj ) {
        return false;
    }
    return true;
}

module.exports = PriorityQueue;

File: src/priorityq.py
"""Create Priority Q object."""


class PriorityQueue(object):
    """Create a new PriorityQueue object."""

    def __init__(self):
        """Initialized value of a Priority Queue."""
        self._priority_dict = {}

    def insert(self, value, priority=0):
        """Insert a new value with priority zero unless otherwise stated."""
        self._priority_dict.setdefault(priority, []).append(value)

    def pop(self):
        """Pop first item in highest priority value."""
        if self._priority_dict == {}:
            raise IndexError('Nothing to pop')
        highest_priority = max(self._priority_dict, key=int)
        popped = self._priority_dict[highest_priority].pop(0)
        if self._priority_dict[highest_priority] == []:
            del self._priority_dict[highest_priority]
        return popped

    def peek(self):
        """Show the first highest priority value."""
        if self._priority_dict == {}:
            return None
        highest_priority = max(self._priority_dict, key=int)
        return self._priority_dict[highest_priority][0]